### DISCUSSION TITLE
Difference between create_state_from_gsd and create_state_from_snapshot?

### DISCUSSION THREAD

**Discussion by tylah25** [glotzerlab/hoomd-blue#1721] [Support] [answered]
What is the difference between the create_state_from_gsd function and create_state_from_snapshot? Besides the obvious differences in what they do (create the simulation state from a GSD file or a snapshot), I've encountered instances where the two functions give me different results. 

For example, if I create an initial configuration file with just one bond type that is not yet present in the system (I will be adding the bonds later with an MD/MC scheme), create_state_from_gsd does not seem to read in this bond type, while create_state_from_snapshot does.

**Comment by joaander** [comment]
There should be no functional difference beyond the obvious. You should find it takes less time to call `create_state_from_gsd` than it does to implement the equivalent snapshot creation in Python.

Regarding the bond types behavior you describe, feel free to open a bug with a minimal test case that shows the problem (i.e. it only has to create a `Simulation`, call `create_state_from_gsd` then print `simulation.state.bond_types`). I would welcome a pull request that fixes the bug. I would guess the root cause is that the types are read inside the `N > 0` check:
https://github.com/glotzerlab/hoomd-blue/blob/37f6348ba60d24ac366411f9b5fe4c8cd0783fc6/hoomd/GSDReader.cc#L251-L288

**Comment by joaander** [comment, marked as answer]
This bug will be fixed in the next release: #1729
